arcgis-ios-sdk-samples/Display information/Create symbol styles from web styles/CreateSymbolStylesFromWebStylesViewController.swift

```diff
@@ -0,0 +1,282 @@
+// Copyright 2021 Esri
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import UIKit
+import ArcGIS
+
+class CreateSymbolStylesFromWebStylesViewController: UIViewController, UIAdaptivePresentationControllerDelegate {
+    // MARK: Storyboard views
+    
+    /// The map view managed by the view controller.
+    @IBOutlet var mapView: AGSMapView! {
+        didSet {
+            mapView.map = makeMap(featureLayer: featureLayer)
+            mapView.setViewpoint(
+                AGSViewpoint(
+                    latitude: 34.28301,
+                    longitude: -118.44186,
+                    scale: 1e4
+                )
+            )
+        }
+    }
+    /// The button to show legend table.
+    @IBOutlet var legendBarButtonItem: UIBarButtonItem!
+    
+    // MARK: Instance properties
+    
+    /// The Esri 2D point symbol style created from a web style.
+    let symbolStyle = AGSSymbolStyle(styleName: "Esri2DPointSymbolsStyle", portal: .arcGISOnline(withLoginRequired: false))
+    /// The observation on the map view's `mapScale` property.
+    var mapScaleObservation: NSKeyValueObservation?
+    /// The data source for the legend table.
+    private var symbolsDataSource: SymbolsDataSource?
+    /// A list of symbols created from the web style, with their associated names and values.
+    private var symbols = [(symbolName: String, symbol: AGSSymbol, symbolCategoryValues: [String])]()
+    /// The legend items for legend table.
+    private var legendItems = [LegendItem]()
```

Comment:
I agree. I was hoping to cram all the required properties into a single struct, to avoid holding stuff piece by piece, such as

```swift
private struct LegendItem {
    let symbolName: String
    let image: UIImage
    let symbol: AGSSymbol
    let symbolCategoryValues: [String]
}
```

But in this case the name of the struct `LegendItem` doesn't seem proper. Do you have any suggestion on this?
Thanks.

Reply:
Yeah, I played around with trying to combine them as well and I didn't like what I was coming up with. But I just realized that the `symbols` property isn't really needed either. It is only read by `makeUniqueValueRenderer(fieldNames:)`, which is called by the completion handler for `getSymbols(symbolStyle:types:completion:)`. You could modify `makeUniqueValueRenderer(fieldNames:)` to take the symbols as an argument. Then, you could either modify the completion handler to pass back the symbols or move the call to `makeUniqueValueRenderer(fieldNames:)` to where the completion handler is called. If you do that, then you could move `legendBarButtonItem.isEnabled = true` to a property observer of `symbolsDataSource` and get rid of the completion handler altogether. Does that make sense?